File: survival-graphics/src/main/java/net/survival/graphics/TextureAtlas.java
package net.survival.graphics;

import java.nio.file.Paths;
import java.util.HashMap;

import net.survival.graphics.opengl.GLFilterMode;
import net.survival.graphics.opengl.GLTexture;
import net.survival.graphics.opengl.GLWrapMode;

public class TextureAtlas implements GraphicsResource
{
    private static final int INITIAL_WIDTH = 64;
    private static final int INITIAL_HEIGHT = 64;

    private final GLTexture textureObject;
    private boolean needsUpdating;

    private BitmapAtlasBuilder atlasBuilder;
    private float inverseWidth;
    private float inverseHeight;
    private HashMap<String, BitmapRegion> texCoords;
    private HashMap<String, Integer> textureWidths;
    private HashMap<String, Integer> textureHeights;

    private final String rootPath;

    public TextureAtlas(String rootPath, boolean mipmappingEnabled) {
        textureObject = new GLTexture();

        if (mipmappingEnabled) {
            textureObject.beginBind()
                    .setMinFilter(GLFilterMode.NEAREST_MIPMAP_NEAREST)
                    .setMagFilter(GLFilterMode.NEAREST)
                    .setWrapS(GLWrapMode.REPEAT)
                    .setWrapT(GLWrapMode.REPEAT)
                    .setMipmapEnabled(true)
                    .setMinLod(0)
                    .setMaxLod(4)
                    .setData(new Bitmap(INITIAL_WIDTH, INITIAL_HEIGHT))
                    .endBind();
        }
        else {
            textureObject.beginBind()
                    .setMinFilter(GLFilterMode.NEAREST)
                    .setMagFilter(GLFilterMode.NEAREST)
                    .setWrapS(GLWrapMode.REPEAT)
                    .setWrapT(GLWrapMode.REPEAT)
                    .setData(new Bitmap(INITIAL_WIDTH, INITIAL_HEIGHT))
                    .endBind();
        }

        needsUpdating = false;

        atlasBuilder = new BitmapAtlasBuilder(INITIAL_WIDTH, INITIAL_HEIGHT);
        inverseWidth = 1.0f / atlasBuilder.getWidth();
        inverseHeight = 1.0f / atlasBuilder.getHeight();

        texCoords = new HashMap<>();
        textureWidths = new HashMap<>();
        textureHeights = new HashMap<>();

        this.rootPath = rootPath;
    }

    @Override
    public void close() {
        textureObject.close();
    }

    public GLTexture getTextureObject() {
        return textureObject;
    }

    public void updateTextures() {
        if (needsUpdating) {
            textureObject.beginBind()
                    .setData(atlasBuilder.build())
                    .endBind();

            needsUpdating = false;
        }
    }

    public int getWidth(String texture) {
        lazyLoadTexture(texture);
        return textureWidths.get(texture);
    }

    public int getHeight(String texture) {
        lazyLoadTexture(texture);
        return textureHeights.get(texture);
    }

    public float getTexCoordU1(String texture) {
        return (float) lazyLoadTexture(texture).getLeft() * inverseWidth;
    }

    public float getTexCoordV1(String texture) {
        return (float) 1.0f - lazyLoadTexture(texture).getTop() * inverseHeight;
    }

    public float getTexCoordU2(String texture) {
        return (float) lazyLoadTexture(texture).getRight() * inverseWidth;
    }

    public float getTexCoordV2(String texture) {
        return (float) 1.0f - lazyLoadTexture(texture).getBottom() * inverseHeight;
    }

    private BitmapRegion lazyLoadTexture(String texture) {
        var region = texCoords.get(texture);
        
        while (region == null) {
            var fullPath = Paths.get(rootPath, texture).toString();
            var bitmap = Bitmap.fromFile(fullPath);
            region = atlasBuilder.addBitmap(bitmap);
            
            if (region == null) {
                rebuildAtlas();
            }
            else {
                texCoords.put(texture, region);
                textureWidths.put(texture, bitmap.getWidth());
                textureHeights.put(texture, bitmap.getHeight());
                needsUpdating = true;
            }
        }
        
        return region;
    }

    private void rebuildAtlas() {
        var newWidth = atlasBuilder.getWidth() * 2;
        var newHeight = atlasBuilder.getHeight() * 2;

        var newAtlasBuilder = new BitmapAtlasBuilder(newWidth, newHeight);
        var newTexCoords = new HashMap<String, BitmapRegion>(texCoords.size());

        for (var entry : texCoords.entrySet()) {
            var texture = entry.getKey();
            var region = entry.getValue();
            var bitmap = region.getBitmap();

            var newRegion = newAtlasBuilder.addBitmap(bitmap);
            newTexCoords.put(texture, newRegion);
        }

        atlasBuilder = newAtlasBuilder;
        inverseWidth = 1.0f / newWidth;
        inverseHeight = 1.0f / newHeight;
        texCoords = newTexCoords;
    }
}

File: survival-graphics/src/main/java/net/survival/graphics/TextureAtlas.kt
package net.survival.graphics

import net.survival.graphics.opengl.GLFilterMode
import net.survival.graphics.opengl.GLTexture
import net.survival.graphics.opengl.GLWrapMode
import java.nio.file.Paths
import java.util.*

private const val INITIAL_WIDTH = 64
private const val INITIAL_HEIGHT = 64

class TextureAtlas(
    private val rootPath: String,
    mipmappingEnabled: Boolean
) {
    val textureObject: GLTexture = GLTexture()
    private var needsUpdating: Boolean
    private var atlasBuilder: BitmapAtlasBuilder
    private var inverseWidth: Float
    private var inverseHeight: Float
    private var texCoords: HashMap<String, BitmapRegion>
    private val textureWidths: HashMap<String, Int>
    private val textureHeights: HashMap<String, Int>

    init {
        if (mipmappingEnabled) {
            textureObject.beginBind()
                .setMinFilter(GLFilterMode.NEAREST_MIPMAP_NEAREST)
                .setMagFilter(GLFilterMode.NEAREST)
                .setWrapS(GLWrapMode.REPEAT)
                .setWrapT(GLWrapMode.REPEAT)
                .setMipmapEnabled(true)
                .setMinLod(0)
                .setMaxLod(4)
                .setData(Bitmap(INITIAL_WIDTH, INITIAL_HEIGHT))
                .endBind()
        }
        else {
            textureObject.beginBind()
                .setMinFilter(GLFilterMode.NEAREST)
                .setMagFilter(GLFilterMode.NEAREST)
                .setWrapS(GLWrapMode.REPEAT)
                .setWrapT(GLWrapMode.REPEAT)
                .setData(Bitmap(INITIAL_WIDTH, INITIAL_HEIGHT))
                .endBind()
        }

        needsUpdating = false

        atlasBuilder = BitmapAtlasBuilder(INITIAL_WIDTH, INITIAL_HEIGHT)
        inverseWidth = 1.0f / atlasBuilder.width
        inverseHeight = 1.0f / atlasBuilder.height

        texCoords = HashMap()
        textureWidths = HashMap()
        textureHeights = HashMap()
    }

    fun close() {
        textureObject.close()
    }

    fun updateTextures() {
        if (needsUpdating) {
            textureObject.beginBind()
                .setData(atlasBuilder.build())
                .endBind()
            needsUpdating = false
        }
    }

    fun getWidth(texture: String): Int {
        lazyLoadTexture(texture)
        return textureWidths[texture]!!
    }

    fun getHeight(texture: String): Int {
        lazyLoadTexture(texture)
        return textureHeights[texture]!!
    }

    fun getTexCoordU1(texture: String): Float =
        lazyLoadTexture(texture).left.toFloat() * inverseWidth

    fun getTexCoordV1(texture: String): Float =
        1.0f - lazyLoadTexture(texture).top * inverseHeight

    fun getTexCoordU2(texture: String): Float =
        lazyLoadTexture(texture).right.toFloat() * inverseWidth

    fun getTexCoordV2(texture: String): Float =
        1.0f - lazyLoadTexture(texture).bottom * inverseHeight

    private fun lazyLoadTexture(texture: String): BitmapRegion {
        var region = texCoords[texture]

        while (region == null) {
            val fullPath = Paths.get(rootPath, texture).toString()
            val bitmap = Bitmap.fromFile(fullPath)
            region = atlasBuilder.addBitmap(bitmap)

            if (region == null) {
                rebuildAtlas()
            }
            else {
                texCoords[texture] = region
                textureWidths[texture] = bitmap.width
                textureHeights[texture] = bitmap.height
                needsUpdating = true
            }
        }

        return region
    }

    private fun rebuildAtlas() {
        val newWidth = atlasBuilder.width * 2
        val newHeight = atlasBuilder.height * 2

        val newAtlasBuilder = BitmapAtlasBuilder(newWidth, newHeight)
        val newTexCoords = HashMap<String, BitmapRegion>(texCoords.size)

        for ((texture, region) in texCoords) {
            val bitmap = region.bitmap
            val newRegion = newAtlasBuilder.addBitmap(bitmap!!)
            newTexCoords[texture] = newRegion!!
        }

        atlasBuilder = newAtlasBuilder
        inverseWidth = 1.0f / newWidth
        inverseHeight = 1.0f / newHeight
        texCoords = newTexCoords
    }
}